daemon/src/process_manager.rs

```diff
@@ -0,0 +1,177 @@
+use crate::db::append_event;
+use crate::model::cfd;
+use crate::model::cfd::CfdEvent;
+use crate::model::cfd::Role;
+use crate::monitor;
+use crate::monitor::MonitorParams;
+use crate::oracle;
+use crate::projection;
+use crate::wallet;
+use anyhow::Context;
+use anyhow::Result;
+use xtra::prelude::MessageChannel;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    db: sqlx::SqlitePool,
+    role: Role,
+    cfds_changed: Box<dyn MessageChannel<projection::CfdsChanged>>,
+    try_broadcast_transaction: Box<dyn MessageChannel<wallet::TryBroadcastTransaction>>,
+    start_monitoring: Box<dyn MessageChannel<monitor::StartMonitoring>>,
+    monitor_collaborative_settlement: Box<dyn MessageChannel<monitor::CollaborativeSettlement>>,
+    monitor_attestation: Box<dyn MessageChannel<oracle::MonitorAttestation>>,
+}
+
+pub struct Event(cfd::Event);
+
+impl Event {
+    pub fn new(event: cfd::Event) -> Self {
+        Self(event)
+    }
+}
+
+impl Actor {
+    pub fn new(
+        db: sqlx::SqlitePool,
+        role: Role,
+        cfds_changed: &(impl MessageChannel<projection::CfdsChanged> + 'static),
+        try_broadcast_transaction: &(impl MessageChannel<wallet::TryBroadcastTransaction> + 'static),
+        start_monitoring: &(impl MessageChannel<monitor::StartMonitoring> + 'static),
+        monitor_collaborative_settlement: &(impl MessageChannel<monitor::CollaborativeSettlement>
+              + 'static),
+        monitor_attestation: &(impl MessageChannel<oracle::MonitorAttestation> + 'static),
+    ) -> Self {
+        Self {
+            db,
+            role,
+            cfds_changed: cfds_changed.clone_channel(),
+            try_broadcast_transaction: try_broadcast_transaction.clone_channel(),
+            start_monitoring: start_monitoring.clone_channel(),
+            monitor_collaborative_settlement: monitor_collaborative_settlement.clone_channel(),
+            monitor_attestation: monitor_attestation.clone_channel(),
+        }
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    fn handle(&mut self, msg: Event) -> Result<()> {
+        let event = msg.0;
+
+        // 1. Safe in DB
+        let mut conn = self.db.acquire().await?;
+        append_event(event.clone(), &mut conn).await?;
+
+        // 2. Post process event
+        match event.event {
+            CfdEvent::ContractSetupCompleted { dlc } => {
+                tracing::info!("Setup complete, publishing on chain now");
+
+                let lock_tx = dlc.lock.0.clone();
+                let txid = self
+                    .try_broadcast_transaction
+                    .send(wallet::TryBroadcastTransaction { tx: lock_tx })
+                    .await??;
```

Comment:
I think we should not fail the entire function here because we will already have saved the event to the DB so this failing should not signal to the caller that they should try again.

Can we extract helpers on `Self` for each command that log on failure and are thus infallible?

Reply:
It can be done, but I think it is dangerous to do this, because we might not notice problems.
I think the proper way is to return different kinds of errors, where some can lead to a retry and some can lead to letting the user know that it failed. But I did not go there yet, needs a bit more thinking. 

I think it is better to fail for now than to try to "cover this up". If we fail the function the user is more likely to notice something is wrong and can e.g. try to restart. Since the event is recorded we should then try to e.g. broadcast again.

Reply:
If we would save more fine-granular events we could record that e.g. publishing a tx failed. I started thinking about that as well, but am not sure about it yet.